pallets/gear/src/lib.rs

```diff
@@ -1436,14 +1443,14 @@ pub mod pallet {
 
                 Self::deposit_event(event);
             } else {
-                // Message in mailbox is not meant for any processing, hence 0 gas limit
-                // and no gas tree needs to be created
-                let origin = who.into_origin();
                 let message = message.into_stored(ProgramId::from_origin(origin));
 
-                // TODO: update logic of insertion into mailbox following new
-                // flow and deposit appropriate event (issue #1010).
-                MailboxOf::<T>::insert(message.clone())?;
+                if gas_limit >= T::MessageRent::get() {
+                    // TODO: update logic of insertion into mailbox following new
+                    // flow and deposit appropriate event (issue #1010).
+                    MailboxOf::<T>::insert(message.clone())?;
+                    let _ = T::GasHandler::create(origin, message_id.into_origin(), gas_limit);
```

Comment:
I want you to provide separate `ValueNode` type, instead of using `External` one. It may be contructed with other `ValueNode::cut` (this is the new one from my imagination), as it does in `split`.
This new type should be able to call only `spend` and `consume` (other variants should return Err), contains origin
Tests should be provided for this value nodes also.
They must be just separated nodes which can only depend on other existing gas in upper tree.

Reply:
1. only `spend` and `consume` means without `split_xx` ( for mutated calls only support `spend` and `consume` )?  
2. what about `ValType::ReservedExternal`?
3. should I provide `ValueNode::cut` in this pr?
4. for a new `ValType` instead of `External`, I may provide a new method for creating this `ValueNode`, mb `T::GasHandler::reserve`

Reply:
> * only `spend` and `consume` means without `split_xx` ( for mutated calls only support `spend` and `consume` )?

Yeah, other mutating methods should always return `Err(_)` with probably something like `Self::error::forbidden()`

> * what about `ValType::ReservedExternal`?

Considering stuff. See the question no.4 answer.

> * should I provide `ValueNode::cut` in this pr?

As I answered in DM, probably the best way is to represent such node in separate PR, review && merge it, and afterward continue this one.

> * for a new `ValType` instead of `External`, I may provide a new method for creating this `ValueNode`, mb `T::GasHandler::reserve`

What if message to mailbox sent without gas specification? You need to depend on some node with value.